Add new reduce options and autograd support for scatter_reduce

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71256
* #70750

Differential Revision: [D33712802](https://our.internmc.facebook.com/intern/diff/D33712802)

mikaylagawarecki committed:
Add new reduce options and autograd support for scatter_reduce

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -1314,21 +1315,47 @@ TORCH_IMPL_FUNC(scatter_reduce_structured_cpu)
+    } else if (reduce == "prod") {
+      out.fill_(1);
+    } else if (reduce == "amax") {
+      out.fill_(INT_MIN);
```
You should be able to use ```std::numeric_limits<scalar_t>::max()```

mikaylagawarecki committed:
Update on "Add new reduce options and autograd support for scatter_reduce"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add new reduce options and autograd support for scatter_reduce"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add new reduce options and autograd support for scatter_reduce"




[ghstack-poisoned]

cpuhrsch reviewed (approved):
Looks great!

mikaylagawarecki committed:
Update on "Add new reduce options and autograd support for scatter_reduce"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add new reduce options and autograd support for scatter_reduce"




[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33712802).

mikaylagawarecki committed:
Update on "Add new reduce options and autograd support for scatter_reduce"


Differential Revision: [D33712802](https://our.internmc.facebook.com/intern/diff/D33712802)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33712802).

mikaylagawarecki committed:
Update on "Add new reduce options and autograd support for scatter_reduce"


Differential Revision: [D33712802](https://our.internmc.facebook.com/intern/diff/D33712802)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33712802).

mikaylagawarecki committed:
fix lint

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33712802).
